lib/ansible/plugins/inventory/__init__.py

```diff
@@ -372,7 +372,7 @@ def _add_host_to_composed_groups(self, groups, variables, host, strict=False):
             self.templar.available_variables = variables
             for group_name in groups:
                 conditional = "{%% if %s %%} True {%% else %%} False {%% endif %%}" % groups[group_name]
-                group_name = original_safe(group_name, force=True)
+                group_name = original_safe(group_name)
```

Comment:
Plugins should be able to override sanitization by setting self._sanitize_group_name, but otherwise the constructed features always sanitize.

This would keep the warning for the change in behavior in 2.8 while letting plugins override:
```diff
diff --git a/lib/ansible/plugins/inventory/__init__.py b/lib/ansible/plugins/inventory/__init__.py
index d34c1e0a0f..9281605298 100644
--- a/lib/ansible/plugins/inventory/__init__.py
+++ b/lib/ansible/plugins/inventory/__init__.py
@@ -377,7 +377,13 @@ class Constructable(object):
             self.templar.available_variables = variables
             for group_name in groups:
                 conditional = "{%% if %s %%} True {%% else %%} False {%% endif %%}" % groups[group_name]
-                group_name = original_safe(group_name, force=True)
+                if self._sanitize_group_name == to_safe_group_name:
+                    # only began to be sanitized in 2.8 - do the same thing as to_safe_group_name but not silently
+                    group_name = original_safe(group_name, force=True)
+                else:
+                    # Allow plugins to do their own sanitization
+                    group_name = self._sanitize_group_name(group_name)
```
On the other hand, maybe the warning isn't needed (there was some confusion because `keyed_groups` sanitized the group names for 2 releases before `groups` began sanitizing) in which case this should be
`group_name = self._sanitize_group_name(group_name)`.

Reply:
Tested your suggested change snippet. It still mangles group names when `ANSIBLE_TRANSFORM_INVALID_GROUP_CHARS=never` is set in environment.

The PR does the fix the bug as-is though.

Reply:
I see the problem, the PR as-is disables the mangling completely. So it does the other opposite of not respecting the `TRANSFORM_INVALID_GROUP_CHARS` configuration.

Reply:
@logan2211  you are incorreclty assuming that this is a toggle for each plugin's validation, it is not, it is for the global validation. Each inventory plugin is free to add it's own validation on top, most do and it is not controlled by that toggle.

This PR alters the behaviour of the plugin which was always meant to filter the invalid names, the plugin itself never allowed - in group names.

Reply:
> the plugin itself never allowed - in group names.

That's not true. Every release prior to 2.8 allowed "-" in group names using the `constructed` plugin, hence #70741. That's why I'm here -- my inventory went from functioning normally, to completely broken when upgrading from 2.7 -> 2.8 due to the group names being arbitrarily mangled by the new behavior.

After reading the docs I thought I found a configuration that allowed disabling this unnecessary mangling behavior, but after finding that it does not restore previous functionality, I ended up opening #70741.

> Each inventory plugin is free to add it's own validation on top, most do and it is not controlled by that toggle.

`constructed` did not mangle group names shown in the #70741 example prior to the change in 2.8 that added the global system. I'd think that the global implementation created a great opportunity for plugins to behave consistently according to a single configuration setting. 

As you said:
https://github.com/ansible/ansible/pull/52748#issuecomment-468017671: "I'm just trying to get the 'general toggle' working with the system w/o breaking any current playbooks"

I'm here because my playbooks are broken. How can we get this fixed?

https://github.com/ansible/ansible/pull/53458#issuecomment-471528717 sums it up very well:
> It is very frustrating to ask for Ansible to not sanitize group names (the current behavior for most things in the current 2.7 version, ec2 is just one case) and have it still replace anything.

Reply:
@logan2211 There are a lot of plugins which also use this code, hence the 'general toggle'. This would support a fix for `constructed` though - the code in `lib/ansible/plugins/inventory/constructed.py` just needs a way to set `self._sanitize_group_name`.

That could be as simple as adding the import `from ansible.inventory.group import to_safe_group_name` and then in parse() add  `self._sanitize_group_name = to_safe_group_name`.

Reply:
> There are a lot of plugins which also use this code, hence the 'general toggle'.

I see that. It looks like it would be fairly easy for most of the plugins to use, rather than ignore the configuration parameter that was added. Simply stop using `force=True` in places like: https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/__init__.py#L43-L45 so the configuration parameter can work as intended.

> This would support a fix for constructed though - the code in lib/ansible/plugins/inventory/constructed.py just needs a way to set self._sanitize_group_name.
> That could be as simple as adding the import from ansible.inventory.group import to_safe_group_name and then in parse() add self._sanitize_group_name = to_safe_group_name.

Thanks, I adjusted #70741 to match your suggested implementation and confirmed that `constructed` now follows the configuration.

Reply:
Also note that implementing your suggestion essentially requires a revert of #60912 because calling `original_safe` directly bypasses `self._sanitize_group_name`, so you'll see this reflected in my PR.